packages/foam-core/src/model/uri.ts

```diff
@@ -0,0 +1,508 @@
+ * ```
+ */
+export interface URI {
+  scheme: string;
+  authority: string;
+  path: string;
+  query: string;
+  fragment: string;
+}
+
+const { posix } = paths;
+const _empty = '';
+const _slash = '/';
+const _regexp = /^(([^:/?#]+?):)?(\/\/([^/?#]*))?([^?#]*)(\?([^#]*))?(#(.*))?/;
+
+export const create = (
+  scheme: string,
+  authority: string,
+  path: string,
+  query: string,
+  fragment: string
+): URI => ({
+  scheme,
+  authority,
+  path,
+  query,
+  fragment,
+});
+
+export const parse = (value: string): URI => {
+  const match = _regexp.exec(value);
+  if (!match) {
+    return create(_empty, _empty, _empty, _empty, _empty);
+  }
+  return create(
+    match[2] || 'file',
+    percentDecode(match[4] || _empty),
+    percentDecode(match[5] || _empty),
+    percentDecode(match[7] || _empty),
+    percentDecode(match[9] || _empty)
+  );
+};
+
+/**
+ * Parses a URI from value, taking into consideration possible relative paths.
+ *
+ * @param reference the URI to use as reference in case value is a relative path
+ * @param value the value to parse for a URI
+ * @returns the URI from the given value. In case of a relative path, the URI will take into account
+ * the reference from which it is computed
+ */
+export const parseWithReference = (value: string, reference: URI): URI => {
+  let uri = parse(value);
+  if (uri.scheme === 'file' && !value.startsWith('/')) {
+    const [path, fragment] = value.split('#');
+    uri = path.length > 0 ? computeRelativeURI(reference, path) : reference;
+    if (fragment) {
+      uri = from(uri, {
+        fragment: fragment,
+      });
+    }
+  }
+  return uri;
+};
+
+export const computeRelativeURI = (
+  reference: URI,
+  relativeSlug: string
+): URI => {
+  // if no extension is provided, use the same extension as the source file
+  const slug =
+    posix.extname(relativeSlug) !== ''
+      ? relativeSlug
+      : `${relativeSlug}${posix.extname(reference.path)}`;
+  return from(reference, {
+    path: posix.join(posix.dirname(reference.path), slug),
+  });
+};
+
+export const file = (path: string): URI => {
+  let authority = _empty;
+
+  // normalize to fwd-slashes on windows,
+  // on other systems bwd-slashes are valid
+  // filename character, eg /f\oo/ba\r.txt
+  if (isWindows) {
+    path = path.replace(/\\/g, _slash);
+  }
+
+  // check for authority as used in UNC shares
+  // or use the path as given
+  if (path[0] === _slash && path[1] === _slash) {
+    const idx = path.indexOf(_slash, 2);
+    if (idx === -1) {
+      authority = path.substring(2);
+      path = _slash;
+    } else {
+      authority = path.substring(2, idx);
+      path = path.substring(idx) || _slash;
+    }
+  }
+
+  return create('file', authority, path, _empty, _empty);
+};
+
+export const placeholder = (key: string): URI => {
+  return {
+    scheme: 'placeholder',
+    authority: '',
+    path: key,
+    query: '',
+    fragment: '',
+  };
+};
+
+export const from = (
```

Comment:
The null/undefined semantics here are a little confusing. As discussed, let's use `create` for the single argument case, and remove the `change` use case.